testcases/kernel/syscalls/recvmmsg/recvmmsg01.c

```diff
@@ -0,0 +1,2250 @@
+		pre += pretv.tv_nsec;
+		unsigned long long pos = postv.tv_sec * 1000 * 1000 * 1000;
+		pos += postv.tv_nsec;
+		test_assert((pos - pre) > (SENDMSG_SLEEP_MS * 1000) / 2);
+	}
+
+	rmcp->rmc_value = val;
+	thread_join(&smcp->smc_thread_call);
+}
+
+/*
+ *	Miscellaneous supporting code is in this section.
+ */
+
+static int errno_value_is_error(int e)
+{
+	return e > 0;  /* negative values are not errors, see non_error_errno */
+}
+
+static void sleep_ms(long ms)
+{
+	test_assert(ms < 1000);
+	struct timespec ts;
+	ts.tv_sec = 0;
+	ts.tv_nsec = ms * 1000 * 1000;
+	nanosleep(&ts, NULL);
+}
+
+static int test_number_get(void)
+{
+	static int test_number_gen;
+
+	/*
+	 * Nested subtests share the same test number, these are multi
+	 * message tests, i.e. recvmmsg() and sendmmsg(), where each message
+	 * is treated as a subtest
+	 */
+
+	if (subtest_number() > 1)
+		return test_number_gen;
+	return ++test_number_gen;
+}
+
+/*
+ * subtests numbers
+ */
+
+static int subtest_level = 1;
+
+static int subtest_number(void)
+{
+	return subtest_level;
+}
+
+static void subtest_nest(void)
+{
+	++subtest_level;
+}
+
+static void subtest_unnest(void)
+{
+	--subtest_level;
+}
+
+static bool subtest_show(void)
+{
+	/*
+	 *	Some tests involve too many subtests, this limits the output.
+	 */
+	return subtest_level <= 20;
+}
+
+/*
+ * compare that memory is equal, makes the output of various ensure() nicer
+ */
+
+static bool mem_is_equal(void *a, void *b, size_t size)
+{
+	return !memcmp(a, b, size);
+}
+
+static bool mem_is_zero(void *a, size_t size)
+{
+	unsigned char set = 0;
+	unsigned char *p = a;
+	unsigned char *endp = p + size;
+	while (p < endp) set |= *p++;
+	return !set;
+}
+
+static void thread_call(struct thread_call *tcp, void *(*func)(void *))
+{
+	SAFE_PTHREAD_CREATE(&tcp->tc_thr, NULL, func, tcp);
+}
+
+static void thread_join(struct thread_call *tcp)
+{
+	SAFE_PTHREAD_JOIN(tcp->tc_thr, NULL);
+}
+
+static void sendmmsg_init(struct sendmmsg_call *smmcp, int sockfd,
+			  unsigned vlen, struct mmsghdr *smm, int flags,
+			  bool call_sendmmsg, bool sender_sleeps)
+{
+	smmcp->smmc_sockfd = sockfd;
+	smmcp->smmc_vlen = vlen;
+	smmcp->smmc_smm = smm;
+	smmcp->smmc_flags = flags;
+	smmcp->smmc_call_sendmmsg = call_sendmmsg;
+	smmcp->smmc_sender_sleeps = sender_sleeps;
+	smmcp->smmc_value = INVALID_SYSCALL_VALUE;
+	smmcp->smmc_errno = 0;
+}
+
+/*
+ * various init and member functions for various types
+ */
+
+static void sockaddr_in_init(struct sockaddr_in *sinp, sa_family_t family,
+			     in_port_t port, uint32_t ip)
+{
+	sockaddr_in_init_to_zero(sinp);
+	sinp->sin_family = family;
+	sinp->sin_port = htons(port);
+	sinp->sin_addr.s_addr = htonl(ip);
+}
+
+static void sockaddr_in_init_to_zero(struct sockaddr_in *sinp)
+{
+	memset(sinp, 0, sizeof(*sinp));
+}
+
+static size_t iovec_max(void)
+{
+	long iov_max = SAFE_SYSCONF(_SC_IOV_MAX);
+	test_assert(iov_max > 2);
+	return (size_t) iov_max;
+}

```

Comment:
I would prefer calling memset directly but if you insist about having a helper for this, all the *_to_zero() function could be better of as a single macro:

#define CLEAR(ptr) memset(ptr, 0, sizeof(*(ptr)))